File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int max = 0;
        int left = 0;
        for(int i = 0; i< s.length();i++) {
            while(!set.add(s.charAt(i)))
                set.remove(s.charAt(left++));
            max = Math.max(max, i - left + 1);
        }

        return max;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # ------- Memory Optimized -------#
        memory = {}
        res = 0
        l = 0
        for r in range(len(s)):
            if s[r] in memory:
                l = max(memory[s[r]] + 1, l)
            memory[s[r]] = r
            res = max(res, r - l + 1)
        return res

        # character set to check character is included
        # visitedCharacters = set()
        # res,l = 0, 0
        # for right in range(len(s)):
        #     while visitedCharacters and s[right] in visitedCharacters:
        #         visitedCharacters.remove(s[l])
        #         l+=1
            
        #     visitedCharacters.add(s[right])
        #     res = max(res, right - l + 1)

        # return res
            
        